### PR TITLE
Adding CloudWatch Logs node + group listing (gated feature)

### PR DESCRIPTION
**Pull request by bryceitoc9**
## Description

First change to get some semblance of CloudWatch Logs functionality in the toolkit! The change does the following:

* Adds a feature flag for CloudWatch Logs
  * Add the following to your VS Code `settings.json` file:
```
    "aws.experimentalFeatureFlags": [
        "CloudWatchLogs"
    ],
```
* With flag added, adds a CloudWatch Logs node to the AWS Explorer (this is pushed into the list and not alphabetized)
* Lists CloudWatch Log Groups for the region when opened
  * Icons are pulled from JetBrains: https://github.com/aws/aws-toolkit-jetbrains/blob/master/jetbrains-core/resources/icons/resources/cloudwatchlogs/CloudWatchLogsGroup_dark.svg and https://github.com/aws/aws-toolkit-jetbrains/blob/master/jetbrains-core/resources/icons/resources/cloudwatchlogs/CloudWatchLogsGroup.svg
* Provides "Copy ARN" functionality for a group as well as an unimplemented "View Log Stream..." action
* Various optimizations to common explorer node code/tests

## Motivation and Context

CloudWatch Logs are cool to have on hand!

## Testing
Tests pass. Functionality empirically works.

## Screenshots (if appropriate)
image

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the changelog using the script `npm run newChange`

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/cloudWatchLogs/explorer/cloudWatchLogsNode.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        updateInPlace(
+            this.logGroupNodes,
+            logGroups.keys(),
+            key => this.logGroupNodes.get(key)!.update(logGroups.get(key)!),
+            key => makeLogGroupNode(this, this.regionCode, logGroups.get(key)!)
+        )
+    }
+}
+
+function makeLogGroupNode(
```
Comment (awschristou) [inline src/cloudWatchLogs/explorer/cloudWatchLogsNode.ts:70]: This method isn't necessary. The contextValue can be set inside LogGroupNode, and the caller of this method can directly call the constructor. Lambda functions did this makeNode approach because those nodes can be created for Lambda parents and CloudFormation Stack parents, which affected how it was created.

File: src/cloudWatchLogs/utils.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import * as vscode from 'vscode'
+import * as nls from 'vscode-nls'
+const localize = nls.loadMessageBundle()
+
+import { CloudWatchLogsClient } from '../shared/clients/cloudWatchLogsClient'
+
+export async function* listCloudWatchLogGroups(
+    client: CloudWatchLogsClient
+): AsyncIterableIterator<CloudWatchLogs.LogGroup> {
+    const status = vscode.window.setStatusBarMessage(
```
Comment (awschristou) [inline src/cloudWatchLogs/utils.ts:16]: I'd prefer to avoid copying this pattern for new code (and address it in existing code separately)...

The problem is if a caller stops iterating before the full result set is returned, the finally block is never entered, meaning the status bar doesn't go away. The status bar would be a concern of the caller.

File: src/extension.ts [hunk 3]
```diff
@@ -225,6 +225,9 @@ function initializeIconPaths(context: vscode.ExtensionContext) {
 
     ext.iconPaths.dark.statemachine = context.asAbsolutePath('resources/dark/stepfunctions/preview.svg')
     ext.iconPaths.light.statemachine = context.asAbsolutePath('resources/light/stepfunctions/preview.svg')
+
+    ext.iconPaths.dark.cloudWatchLogGroup = context.asAbsolutePath('resources/dark/log-group.svg')
+    ext.iconPaths.light.cloudWatchLogGroup = context.asAbsolutePath('resources/dark/log-group.svg')
```
Comment (awschristou) [inline src/extension.ts:230]: copy paste issue (dark svg referenced)

File: src/test/lambda/explorer/lambdaFunctionNode.test.ts [hunk 4]
```diff
@@ -4,26 +4,26 @@
...
+import { CloudWatchLogs } from 'aws-sdk'
 import * as os from 'os'
-import { LambdaFunctionNode } from '../../../lambda/explorer/lambdaFunctionNode'
+import { LogGroupNode } from '../../../CloudWatchLogs/explorer/logGroupNode'
 import { ext } from '../../../shared/extensionGlobals'
 import { TestAWSTreeNode } from '../../shared/treeview/nodes/testAWSTreeNode'
 import { clearTestIconPaths, IconPath, setupTestIconPaths } from '../../shared/utilities/iconPathUtils'
 
-describe('LambdaFunctionNode', () => {
+describe('LogGroupNode', () => {
```
Comment (awschristou) [inline src/test/lambda/explorer/lambdaFunctionNode.test.ts:14]: Why are these edits in the lambda function node test file?

File: src/cloudWatchLogs/utils.ts (hunk 2 above)
Comment (bryceitoc9) [reply src/cloudWatchLogs/utils.ts:16]: Interesting, any proposed ways around that? I'm assuming we don't want to drop the status bar?

File: src/awsexplorer/activation.ts [hunk 5]
```diff
@@ -197,6 +198,12 @@ async function registerAwsExplorerCommands(
             }
         )
     )
+
+    context.subscriptions.push(
+        vscode.commands.registerCommand('aws.viewLogStream', async (node: LogGroupNode) =>
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/awsexplorer/activation.ts:203]: Maybe `aws.cloudWatchLogs.viewLogStream`? These commands shouldn't necessarily be exclusive to the explorer (I had a thought at one point to allow this through the command palette, for instance).

Do you think this individual change is one that should be made here or should that wait for the full pass?

File: src/cloudWatchLogs/utils.ts (hunk 2 above)
Comment (awschristou) [reply src/cloudWatchLogs/utils.ts:16]: We might be able to start without the status bars. The Tree nodes already have a loading indicator while they are getting data.

File: src/awsexplorer/activation.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (awschristou) [reply src/awsexplorer/activation.ts:203]: > These commands shouldn't necessarily be exclusive to the explorer

This command operates on an explorer node (LogGroupNode), it seems likely there would be a similar command for non-explorer situations.

> Do you think this individual change is one that should be made here or should that wait for the full pass?

It would be good to start introducing code that we can use as reference to future implementers instead of propagating patterns we believe should change. I'll defer to your judgement on this one.

[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/awsexplorer/activation.ts:203]: > > These commands shouldn't necessarily be exclusive to the explorer
> 
> This command operates on an explorer node (LogGroupNode), it seems likely there would be a similar command for non-explorer situations.

The ideal here would be something like:
```
async (node?: LogGroupNode) {
    let logGroup = logGroupNode?.name
    if (!logGroup) {
        logGroup = # workflow to select a log group name
    }
    # continue workflow
}
```

I'd be interested to see if that's a pattern that we'd want going forward so we can have additional ways into each workflow.

> > Do you think this individual change is one that should be made here or should that wait for the full pass?
> 
> It would be good to start introducing code that we can use as reference to future implementers instead of propagating patterns we believe should change. I'll defer to your judgement on this one.

Be the change you want to see in the world, I guess. Coming right up.

File: src/test/lambda/explorer/lambdaFunctionNode.test.ts (hunk 4 above)
Comment (bryceitoc9) [reply src/test/lambda/explorer/lambdaFunctionNode.test.ts]: Because I copy/pasted and edited the source instead of the copy -_-

File: src/shared/clients/defaultCloudWatchLogsClient.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+import { CloudWatchLogs } from 'aws-sdk'
+import { ext } from '../extensionGlobals'
+import { CloudWatchLogsClient } from './cloudWatchLogsClient'
+
+export class DefaultCloudWatchLogsClient implements CloudWatchLogsClient {
+    public constructor(public readonly regionCode: string) {}
+
+    public async *describeLogGroups(): AsyncIterableIterator<CloudWatchLogs.LogGroup> {
```
Comment (ckant) [inline src/shared/clients/defaultCloudWatchLogsClient.ts:13]: I saw this approach in another impl. I don't think what it's doing here is actually desired behavior.

Consider the usage of this. It loads the entire next page on demand as soon as it's needed. 

On the one hand, the caller is completely unaware of this service call. This makes it easier for them to use this.

On the other hand, the caller *should* be aware of the service call. The normal use case would be to coordinate the pagination with a "Load more..." action. The caller is then completely aware of exactly what they have and exactly when a service call is made.

Basically in order to use an async iterable here, you would return an async interable of *pages of log groups* rather than an async interable of *log groups*.

That gives the caller control over pagination without them needing to keep track of the token.
That being said, I honestly think the caller is better off just knowing about the continuation token. It avoids the caller unintentionally treating iteration as a "cheap" operation when it actually results in a service call.

File: src/shared/clients/defaultCloudWatchLogsClient.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    protected async invokeDescribeLogGroups(
+        request: CloudWatchLogs.DescribeLogGroupsRequest,
+        sdkClient: CloudWatchLogs
+    ): Promise<CloudWatchLogs.DescribeLogGroupsResponse> {
+        return sdkClient.describeLogGroups(request).promise()
+    }
+
+    protected async createSdkClient(): Promise<CloudWatchLogs> {
```
Comment (ckant) [inline src/shared/clients/defaultCloudWatchLogsClient.ts:32]: Consider providing the SDK client in the constructor. It lets you mock the Cloudwatch impl without having to change global state via `ext.sdkClientBuilder`.

File: src/shared/featureToggle.ts [hunk 8]
```diff
@@ -3,15 +3,20 @@
...
-export enum ActiveFeatureKeys {}
+/**
+ * This enum represents the list of active feature flags to check for
+ * Features all need to be added with matching enums and string values
+ * e.g. `Feature1 = 'Feature1'`
+ * Features that do not follow this scheme will not work.
+ * You cannot have more active features than FeatureToggle.maxFeatures (default: 5)
+ * Any features that are flagged in the code but not added here will always return false.
+ */
+export enum ActiveFeatureKeys {
```
Comment (ckant) [inline src/shared/featureToggle.ts:17]: Do we use enums elsewhere? I think we should stick to union types.

```typescript
export type ActiveFeatureKeys = 'CloudWatchLogs' | '...' | '...'
```

From what I can see, enums are generally avoided in typescript.

File: src/shared/featureToggle.ts [hunk 9]
```diff
@@ -44,7 +49,12 @@ export class FeatureToggle {
 
     private readonly maxFeatures: number = 5
 
-    public constructor(configuration: SettingsConfiguration, overrideKeys?: string[]) {
+    private static INSTANCE: FeatureToggle | undefined
+
+    public constructor(
```
Comment (ckant) [inline src/shared/featureToggle.ts:54]: If you're going for a lazily loaded singleton, is there a reason that the constructor must also be public? Multiple instances could escape here.

File: src/test/shared/clients/mockClients.ts [hunk 10]
```diff
@@ -16,49 +17,63 @@ import { ToolkitClientBuilder } from '../../../shared/clients/toolkitClientBuild
...
-        private readonly schemaClient: SchemaClient = new MockSchemaClient(),
-
-        private readonly ecsClient: EcsClient = new MockEcsClient({}),
-
-        private readonly iamClient: IamClient = new MockIamClient({}),
-
-        private readonly lambdaClient: LambdaClient = new MockLambdaClient({}),
-
-        private readonly stepFunctionsClient: StepFunctionsClient = new MockStepFunctionsClient(),
+interface Clients {
```
Comment (ckant) [inline src/test/shared/clients/mockClients.ts:20]: Best to export interfaces that are accepted as part of another exported type.

File: src/awsexplorer/regionNode.ts [hunk 11]
```diff
@@ -6,12 +6,25 @@
...
 import { Region } from '../shared/regions/endpoints'
 import { RegionProvider } from '../shared/regions/regionProvider'
 import { AWSTreeNodeBase } from '../shared/treeview/nodes/awsTreeNodeBase'
 import { StepFunctionsNode } from '../stepFunctions/explorer/stepFunctionsNodes'
 
+/**
+ * Services that are candidates to add to the region explorer.
+ * `serviceId`s are checked against ~/resources/endpoints.json to see whether or not the service is available in the given region.
+ * If the service is available, we use the `createFn` to generate the node for the region.
+ * This interface exists so we can add additional nodes to the array (otherwise Typescript types the array to what's already in the array at creation)
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/awsexplorer/regionNode.ts:21]: I'll see what I can do. I do like having some degree of documentation for what's going into the object though.

File: src/cloudWatchLogs/explorer/cloudWatchLogsNode.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public async updateChildren(): Promise<void> {
+        const client: CloudWatchLogsClient = ext.toolkitClientBuilder.createCloudWatchLogsClient(this.regionCode)
+        const logGroups: Map<string, CloudWatchLogs.LogGroup> = toMap(
+            await toArrayAsync(client.describeLogGroups()),
+            configuration => configuration.logGroupName
+        )
+
+        updateInPlace(
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/cloudWatchLogs/explorer/cloudWatchLogsNode.ts:58]: I'm not entirely sure if there was a specific reason that we opted for this process. Thoughts, @awschristou ?

File: src/cloudWatchLogs/explorer/logGroupNode.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        this.iconPath = {
+            dark: Uri.file(ext.iconPaths.dark.cloudWatchLogGroup),
+            light: Uri.file(ext.iconPaths.light.cloudWatchLogGroup),
+        }
+        this.contextValue = CONTEXT_VALUE_CLOUDWATCH_LOG
+    }
+
+    public update(configuration: CloudWatchLogs.LogGroup): void {
+        this.configuration = configuration
+        this.label = this.configuration.logGroupName || ''
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/cloudWatchLogs/explorer/logGroupNode.ts:32]: No idea. The type allows it though. Nothing would break if this returned `''`, at least in the tree view (the to-be-added `View Log Streams...` command may though)

File: src/cloudWatchLogs/explorer/logGroupNode.ts [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        this.tooltip = `${this.configuration.logGroupName}${os.EOL}${this.configuration.arn}`
+    }
+
+    public get logGroupName(): string {
+        return this.configuration.logGroupName || ''
+    }
+
+    public getArn(): string {
+        return this.configuration.arn || ''
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/cloudWatchLogs/explorer/logGroupNode.ts:41]: Yup.

File: src/shared/clients/defaultCloudWatchLogsClient.ts (hunk 7 above)
Comment (bryceitoc9) [reply src/shared/clients/defaultCloudWatchLogsClient.ts:32]: Interesting. Mind if that waits for a future PR? Would like to one-fell-swoop that across other clients.

File: src/shared/clients/defaultCloudWatchLogsClient.ts (hunk 6 above)
Comment (bryceitoc9) [reply src/shared/clients/defaultCloudWatchLogsClient.ts:13]: I think I'll hold off on this change for now. Ultimately, unless we want a "Load More..." button, the caller is going to essentially implement the same do-while loop, which I guess makes the caller more conscious but at the same time we're pretty much using these calls at this time on a 1:1 basis.

File: package.json [hunk 15]
```diff
@@ -596,6 +596,11 @@
                     "command": "aws.hideRegion",
                     "group": "0@1",
                     "when": "view == aws.explorer && viewItem == awsRegionNode"
+                },
+                {
+                    "command": "aws.cloudWatchLogs.viewLogStream",
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply package.json:601]: Would be nice but can be easily added later. I'll deem it out-of-scope for this PR.

File: src/cloudWatchLogs/explorer/cloudWatchLogsNode.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (awschristou) [reply src/cloudWatchLogs/explorer/cloudWatchLogsNode.ts:58]: The current pattern isn't ideal, it is compensating for not using the TreeItem.Id property, to keep positioning and expanded state stable across refreshes. That should probably be a wholesale pass to improve node-child code

File: src/test/cloudWatchLogs/explorer/logGropNode.test.ts [hunk 16]
```diff
@@ -1,29 +1,29 @@
 /*!
```
Comment (awschristou) [inline src/test/cloudWatchLogs/explorer/logGropNode.test.ts:1]: filename has a typo (grop)

File: src/shared/featureToggle.ts (hunk 8 above)
Comment (justinmk3) [reply src/shared/featureToggle.ts:17]: > We do use enums in a few other places. I think I opted for enums because I liked the simplicity of specifying `ActiveFeatureKeys.CloudWatchLogs` etc. elsewhere.

Also we end up defining constants for the union values, which is a bit redundant...

File: src/test/cloudWatchLogs/explorer/logGropNode.test.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    it('initializes the label', async () => {
+        assert.strictEqual(testNode.label, fakeLogGroup.logGroupName)
+    })
+
+    it('initializes the functionName', async () => {
+        assert.strictEqual(testNode.logGroupName, fakeLogGroup.logGroupName)
+    })
+
+    it('initializes the tooltip', async () => {
+        assert.strictEqual(testNode.tooltip, `${fakeLogGroup.logGroupName}${os.EOL}${fakeLogGroup.arn}`)
```
Comment (justinmk3) [inline src/test/cloudWatchLogs/explorer/logGropNode.test.ts:54]: concept of "the widget initializes itself" is trivial, so I would put all of these tests in 1 test.

```
it('initializes properties'), async() => {
  assert.strictEqual(testNode.regionCode, 'someregioncode')
  assert.strictEqual(testNode.label, fakeLogGroup.logGroupName)
  ...
})
```

Comment (bryceitoc9) [reply src/test/cloudWatchLogs/explorer/logGropNode.test.ts]: This is a copy/paste of some of the other resource node tests. I'll leave it out of this so we can do a full pass on these later (I'd also like to explore cutting down on the repeated code, I'll leave that for another day).

File: src/awsexplorer/regionNode.ts [hunk 18]
```diff
@@ -35,13 +48,19 @@ export class RegionNode extends AWSTreeNodeBase {
...
+        const serviceCandidates: ServiceCandidate[] = [
             { serviceId: 'cloudformation', createFn: () => new CloudFormationNode(this.regionCode) },
             { serviceId: 'lambda', createFn: () => new LambdaNode(this.regionCode) },
             { serviceId: 'schemas', createFn: () => new SchemasNode(this.regionCode) },
             { serviceId: 'states', createFn: () => new StepFunctionsNode(this.regionCode) },
         ]
 
+        // Feature Toggle for CloudWatch Logs
+        // REMOVE_WHEN_CLOUDWATCH_LOGS_READY
+        if (FeatureToggle.getFeatureToggle().isFeatureActive(ActiveFeatureKeys.CloudWatchLogs)) {
```
[the remark this answers is not in the document]
Comment (ckant) [reply src/awsexplorer/regionNode.ts]: If we're worried about too many dependencies for extension state stuff, couldn't we just roll them into one object and pass that one around?

This is just one example of global state. There's many places in the code base where people ended up adding global stuff they want to control/test to the constructor (e.g. telemetry).

It's quite a bit nicer to provide a fake this way versus having to set up and tear down global state in tests.

File: src/cloudWatchLogs/explorer/cloudWatchLogsNode.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (ckant) [reply src/cloudWatchLogs/explorer/cloudWatchLogsNode.ts:58]: Ah, VSCode actually (now?) maintains the expanded state across refreshes. It's got an internal representation of all of the nodes and looks like it uses the hierarchy + label to figure out a new node is the same as an existing node.

File: src/shared/clients/defaultCloudWatchLogsClient.ts (hunk 6 above)
Comment (ckant) [reply src/shared/clients/defaultCloudWatchLogsClient.ts:13]: If that's the case, could you change this to loop through all results and return an array? Makes the impl simpler and it's easier and less surprising for the caller.

File: src/shared/featureToggle.ts (hunk 8 above)
Comment (ckant) [reply src/shared/featureToggle.ts:17]: Yeah it's weird because the type unions give you basically the same value.

You get autocomplete and compile-time type checking. There's not really a reason to define constants for the items in the union. The strings *are* the constants. You just can't do `functionThatAcceptsEnum(MyEnum.<autocomplete>)` you just do `functionThatAcceptsUnion('<autocomplete>`) instead (i.e. start typing a string).

Either way works though.

Comment (bryceitoc9) [reply src/shared/featureToggle.ts:17]: From an offline conversation: I don't think union types let you programmatically access the types (via something like `Object.keys`), so you can't build a map off of them
